### PR TITLE
Custom error response handling

### PR DESCRIPTION
**Pull request by kellymclaughlin**
This is a change that stems from the discussion on #93. That PR was an attempt to fix a bug with returning a custom response body for requests for unmatched routes, but this PR represents a more broad effort to improve handling of requests for unmatched routes and also provide a means to return a custom response body for any error code. 

~~_Note:_ This work is currently incomplete. I have only added handling for the custom responses at `l07` in the decision tree. I'm opening the PR now to garner feedback on the approach before going any further.~~

Here is a short summary of these changes mostly taken from the commit message:
- Do not traverse the decision tree when no route matches. This mimics
  the [behavior of Erlang Webmachine](https://github.com/webmachine/webmachine/blob/master/src/webmachine_mochiweb.erl#L62-L64) and simplifies dealing with such
  requests. 
- Add the ability to specify a mapping of HTTP status codes to pairs of
  content type header values and custom response bodies for an Airship
  application or for an individual resource. The content type value in
  the mapping is used to set the response `Content-Type` header and the
  intention is that the specified content type value match the error
  response body.In the case of conflicts between application-level and
  resource-level error response mappings, the resource-level mappings
  are favored. This also more closely mimics what is offered by Erlang Webmachine.
- Replace the `Resource IO` parameter to `resourceToWai` with an
  ErrorResponses type that represents the mapping from HTTP status code
  to error response data. The 404 resource previously used to handle
  unmatched routes is no longer necessary since requests for unmatched
  routes now bail out early.

### CODE REVIEW

**Comment by charleso** [comment]
@kellymclaughlin Despite my last, slightly confusing comment on your other PR this seems like a nice approach, and doesn't overload the `contentTypesProvided` field.

A few questions:
- The `type ErrorResponses = Map HTTP.Status (ByteString, ResponseBody)` seems a little restricted. Would it be expected to be able to handle different `Accept` types? Something like `Map HTTP.Status [NonEmpty (ByteString, ResponseBody)]` perhaps?
- I vaguely wonder whether you would want to be wrap the tuple in `m`? For example, would there be anything in the common `StateT s` that would be important/required in some cases?

Looks good though. :)

File: src/Airship/Internal/Helpers.hs [hunk 1]
```diff
@@ -94,21 +95,40 @@ toWaiResponse Response{..} cfg trace quip =
...
+            let (ct, respBody) = fromMaybe (mempty, Empty) response404
+                resp = Response HTTP.status404 [(HTTP.hContentType, ct)] respBody
+            respond =<< (return $ toWaiResponse resp cfg "" quip)
+        Just (resource, (params', matched)) -> do
+            nowTime <- getCurrentTime
+            -- Merge the application-wide error responses with the resource
+            -- specific error responses favoring the resource-specific
+            --in case of conflict
+            let mergedResource = resource { errorResponses = M.union (errorResponses resource) errors }
+            (=<<) respond . run req $ do
```
Comment (patrickt) [inline src/Airship/Internal/Helpers.hs]: This (the prefix-applied `bind`) is sort of an odd idiom here; I was able to decipher it after a while, but I'd avoid the one-liner here.

File: src/Airship/Internal/Route.hs [hunk 2]
```diff
@@ -80,14 +80,19 @@ star = Route [RestUnbound]
 newtype RoutingSpec m a = RoutingSpec { getRouter :: Writer [(Route, Resource m)] a }
     deriving (Functor, Applicative, Monad, MonadWriter [(Route, Resource m)])
 
-route :: [(Route, a)] -> [Text] -> a -> (a, (HashMap Text Text, [Text]))
-route routes pInfo resource404 = foldr' (matchRoute pInfo) (resource404, (mempty, mempty)) routes
+route :: [(Route, a)]
+      -> [Text]
+      -> Maybe (a, (HashMap Text Text, [Text]))
```
Comment (patrickt) [inline src/Airship/Internal/Route.hs]: When I see doubly-nested tuple types—note that we have an `(a, (b, [c]))` structure here, all wrapped in a Maybe—I find it hard to juggle the types in my head. Can we use a data structure for these various types, or, failing that, rub some `type` aliases on them? At the very least, we can destructure a couple of those tuples.

Comment (patrickt) [inline src/Airship/Internal/Route.hs]: There also might be more expressive alternatives to folds we can think of to build routing primitives. `mapAccumL` is a good one. This could all be addressed in another PR though.

File: src/Airship/Resource.hs [hunk 3]
```diff
@@ -141,4 +142,5 @@ defaultResource = Resource { allowMissingPost          = return False
                            , serviceAvailable          = return True
                            , uriTooLong                = return False
                            , validContentHeaders       = return True
+                           , errorResponses            = empty
```
Comment (patrickt) [inline src/Airship/Resource.hs]: ultra-pedantry: use `mempty` here and you won't have to use that import line

File: src/Airship/Types.hs [hunk 4]
```diff
@@ -101,19 +104,21 @@ data ResponseBody
...
+data Response = Response { _responseStatus  :: Status
+                         , _responseHeaders :: ResponseHeaders
+                         , _responseBody    :: ResponseBody
                          }
 
-data ResponseState = ResponseState { stateHeaders   :: ResponseHeaders
-                                   , stateBody      :: ResponseBody
-                                   , _params        :: HashMap Text Text
-                                   , _dispatchPath  :: [Text]
+data ResponseState = ResponseState { stateHeaders  :: ResponseHeaders
```
Comment (patrickt) [inline src/Airship/Types.hs]: While you're here, mind fixing the inconsistent underscore convention we have in this file? I think we added `_params` and `_dispatchPath` after we dropped in microlens, so now that we have lenses let's just use them everywhere.

File: src/Airship/Types.hs [hunk 5]
```diff
@@ -101,19 +104,21 @@ data ResponseBody
...
+                                   , stateBody     :: ResponseBody
+                                   , _params       :: HashMap Text Text
+                                   , _dispatchPath :: [Text]
                                    }
 
 type Trace = [Text]
 
+type ErrorResponses = Map HTTP.Status (ByteString, ResponseBody)
```
Comment (patrickt) [inline src/Airship/Types.hs]: Nitpick: I would actually use a singular noun for this type alias. `ErrorRecord`? `FailedResponseRecord`? Ehh. I'm failing to convince myself, really. Whatever.

Comment (patrickt) [inline src/Airship/Types.hs]: Also can we use a `HashMap` here instead of a Map? Would kill our extra `containers` dependency, though not by much.

File: src/Airship/Internal/Decision.hs [hunk 6]
```diff
@@ -564,7 +565,13 @@ l07 r = do
     req <- lift request
     if requestMethod req == HTTP.methodPost
         then m07 r
-        else lift $ halt HTTP.status404
+        else do
```
Comment (patrickt) [inline src/Airship/Internal/Decision.hs]: can you use one `lift` here instead of the three later on?

File: src/Airship/Internal/Helpers.hs [hunk 7]
```diff
@@ -94,21 +95,40 @@ toWaiResponse Response{..} cfg trace quip =
...
+        response404 = M.lookup HTTP.status404 errors
     quip <- getQuip
-    (=<<) respond . run req $ do
-      (response, trace) <- eitherResponse nowTime params' matched req (flow resource)
-      return $ toWaiResponse response cfg (traceHeader trace) quip
+    case route routeMapping pInfo of
+        Nothing -> do
+            -- No route matched so bail out
+            let (ct, respBody) = fromMaybe (mempty, Empty) response404
+                resp = Response HTTP.status404 [(HTTP.hContentType, ct)] respBody
```
Comment (patrickt) [inline src/Airship/Internal/Helpers.hs]: weird pedantry, but can we add another `let` there? I know it hangs over from the previous `let`, but the indentation confused me.

File: airship.cabal [hunk 8]
```diff
@@ -42,6 +42,7 @@ library
                       , bytestring
                       , bytestring-trie == 0.2.4.*
                       , case-insensitive
+                      , containers
```
Comment (patrickt) [inline airship.cabal:45]: see my later comments on wondering if we can get away with a HashMap

**Comment by tmcgilchrist** [comment]
Nice work, I do think you'd want access to the state when generating the response for errors like @charleso said.

**Comment by kellymclaughlin** [comment]
@charleso Thanks for those questions and feedback. For your first question, I had the handling the differen `Accept` types in my mind for this, but wasn't sure if that would be useful, but since you bring it up as well I think I will add that in. For the 2nd question about state access, it seems like a good idea so I'll gladly incorporate that.

Thanks also for the feedback @patrickt and @tmcgilchrist! I'll work on incorporating this feedback and completing the changes to the decision module and reping this PR when I have it finished.

**Comment by kellymclaughlin** [comment]
Ok, I've completed a reworking of the original idea based on the previous feedback. The content-type of the error response for a given status code is now negotiated based on the request's Accept header value. The type of the error responses that users specify is now `Webmachine m ResponseBody`. As an example I modified the Basic example to illustrate accessing state and performing an IO action on a 404:

```
main :: IO ()
main = do
    static <- staticResource Cache "assets"
    let port = 3000
        host = "127.0.0.1"
        settings = setPort port (setHost host defaultSettings)
        routes = myRoutes static
    mvar <- newMVar HM.empty
    let s = State mvar
    putStrLn "Listening on port 3000"
    runSettings settings (resourceToWaiT defaultAirshipConfig (const $ flip evalStateT s) routes errors)
    where
        errors :: (MonadIO m, MonadState State m) => M.Map HTTP.Status [(MediaType, Webmachine m ResponseBody)]
        errors = let _response404 = escapedResponse "<html><head></head><body><h1>404 Not Found</h1></body></html>"
                 in M.singleton HTTP.status404 [("text/html", response404Action)]

response404Action :: (MonadIO m, MonadState State m) => Webmachine m ResponseBody
response404Action = do
    s <- lift get
    _m <- liftIO (readMVar (_getState s))
    liftIO $ print ("Hooray, IO!" :: String)
    return $ escapedResponse "<html><head></head><body><h1>404 Not Found</h1></body></html>"
```

It's a contrived example, but it shows that there could be a lot of useful possibilities for customizing an application or resource. I have updated all calls to `halt` for 4XX or 5XX in the Decision module to instead be calls to `haltWithErrorResponse`. 

Patrick had previously asked about using `HashMap` instead of `Map` to avoid bringing in the `containers` package. The reason I went with `Map` is that the `Network.HTTP.Types.Status` type does not have a `Hashable` instance so instead of going with just integers to represent the status codes I used `Map`.

File: src/Airship/Internal/Decision.hs [hunk 9]
```diff
@@ -131,6 +131,32 @@ writeCacheTags Resource{..} = lift $ do
...
+-- | optionally set based on the error responses provided by the application and resource.
+-- | The contents of the 'Webmachine''s response body will be streamed back to the client.
+-- | This is a shortcut for constructing a 'Response' with 'getResponseHeaders' and 'getResponseBody'
+-- | and passing that response to 'finishWith'.
+haltWithErrorResponse :: Monad m => HTTP.Status -> Resource m -> Webmachine m a
+haltWithErrorResponse status Resource{..} = do
+    req <- request
+    let reqHeaders = requestHeaders req
+        acceptStr = lookup HTTP.hAccept reqHeaders
+        errBodies = map dupContentType <$> M.lookup HTTP.status404 errorResponses
```
Comment (charleso) [inline src/Airship/Internal/Decision.hs]: I'm assuming the `status404` is a bug here and should be `status`?

File: src/Airship/Internal/Decision.hs [hunk 10]
```diff
@@ -131,6 +131,32 @@ writeCacheTags Resource{..} = lift $ do
...
+        mResp = join $ mapAcceptMedia <$> errBodies <*> acceptStr
+    when (isJust mResp) $ do
+       let (Just (ct, body)) = mResp
```
Comment (charleso) [inline src/Airship/Internal/Decision.hs]: Would suggest pattern matching or `forM_` here rather than the `isJust` and unsafe  pattern matching

File: src/Airship/Types.hs [hunk 11]
```diff
@@ -217,6 +228,29 @@ eitherResponse reqDate reqParams dispatched req resource = do
...
+    let reqHeaders = requestHeaders req
+        acceptStr = lookup HTTP.hAccept reqHeaders
+        dupContentType (a, b) = (a, (a, b))
+        errBodies = fmap dupContentType <$> M.lookup HTTP.status404 errors
+        mResp = join $ mapAcceptMedia <$> errBodies <*> acceptStr
+        (ct, respBody) = case mResp of
+            (Just (c, b)) -> do
+                (c, b)
+            Nothing ->
+                ("text" // "plain", return Empty)
```
Comment (charleso) [inline src/Airship/Types.hs]: `fromMaybe ("text" // "plain", return Empty) mResp` /codegolf

File: src/Airship/Types.hs [hunk 12]
```diff
@@ -217,6 +228,29 @@ eitherResponse reqDate reqParams dispatched req resource = do
     (e, trace) <- runWebmachine reqDate reqParams dispatched req resource
     return (both e, trace)
 
+errorResponse :: Monad m =>
+                 UTCTime
+              -> Request
+              -> ErrorResponses m
+              -> m Response
```
Comment (charleso) [inline src/Airship/Types.hs]: I'm not sure but I suspect if you make this function return `Webmachine m Response` you can remove some of the duplication with `resourceToWaitT` (eg. `runRWST`) down below.

File: src/Airship/Internal/Helpers.hs (hunk 1 above)
Comment (charleso) [inline src/Airship/Internal/Helpers.hs]: > This (the prefix-applied bind) is sort of an odd idiom here;

Apologies, that was me, it's a bad habit of mine to avoid coming up with variable names.

File: src/Airship/Internal/Decision.hs [hunk 13]
```diff
@@ -131,6 +131,32 @@ writeCacheTags Resource{..} = lift $ do
        Nothing -> return ()
        Just d  -> addResponseHeader ("Last-Modified", utcTimeToRfc1123 d)
 
+-- | Immediately halts processing with the provided 'Status' code. The response body is
+-- | optionally set based on the error responses provided by the application and resource.
+-- | The contents of the 'Webmachine''s response body will be streamed back to the client.
+-- | This is a shortcut for constructing a 'Response' with 'getResponseHeaders' and 'getResponseBody'
+-- | and passing that response to 'finishWith'.
+haltWithErrorResponse :: Monad m => HTTP.Status -> Resource m -> Webmachine m a
+haltWithErrorResponse status Resource{..} = do
```
Comment (charleso) [inline src/Airship/Internal/Decision.hs]: This suggestion might be a pain.

One small problem with having something different to `halt` is that consumers will not get this functionally for free (if they're naughty and use `halt` that is), as well as you having to change all of the `Decision.hs` functions manually.

One suggestion which you should feel _completely_ free to ignore would be to do (something like) inspect the `Response` in `resourceToWaiT` and if no body is set call `errorResponse`. The danger I see is if you really really want to halt with a status but no body and also have an error handler for the same status. Just an idea, I'm curious if anyone has an opinion on that one.

**Comment by charleso** [comment]
Looking good @kellymclaughlin

File: src/Airship/Internal/Decision.hs (hunk 13 above)
Comment (kellymclaughlin) [inline src/Airship/Internal/Decision.hs]: Yes, I had a similar thought about whether or not this was a case that my be of interest/concern to users. I like idea of checking for a response body and only falling back to the resource's default error responses if it is not set. 

The case where someone has defined an error response for a particular status code and then wants to return a response with no body at some point in a resource seems to me like enough of an edge case that I can live with making the path to accomplish that to not define an error response for that status code for the resource and then the user must set the response body manually at the points in their resource code where they would like it returned. 

I'll do some playing around with this and see what comes of it.

File: src/Airship/Internal/Decision.hs (hunk 9 above)
Comment (kellymclaughlin) [inline src/Airship/Internal/Decision.hs]: Good catch, definitely a bug. Will fix.

**Comment by kellymclaughlin** [comment]
Reworked the approach is based on [feedback](https://github.com/helium/airship/pull/94#discussion-diff-56925630) from @charleso. This adjusted approach yields a much smaller set of changes that still achieves the same results.

File: src/Airship/Types.hs [hunk 14]
```diff
@@ -140,6 +147,7 @@ instance MonadBaseControl b m => MonadBaseControl b (Webmachine m) where
 -- | A convenience synonym that writes the @Monad@ type constraint for you.
 type Handler m a = Monad m => Webmachine m a
```
Comment (tmcgilchrist) [inline src/Airship/Types.hs]: Hmm, thought we got rid of this synonym?

**Comment by tmcgilchrist** [comment]
Less changes for the same behaviour is a win. :rocket: 
I find the `resourceToWai` and `errorResponse` functions more complicated than I'd like but that's just personal preference.

This is good by me :+1:

File: src/Airship/Types.hs (hunk 14 above)
Comment (reiddraper) [inline src/Airship/Types.hs]: Looks like it snuck back in, in 3c97d750.

Comment (reiddraper) [inline src/Airship/Types.hs]: Or rather, was never removed?

Comment (tmcgilchrist) [inline src/Airship/Types.hs]: Might be remembering we removed it on our branch.

I don't think it should be there, but we have had that conversation before :-) . I'm curious if you're using it anywhere?

Comment (reiddraper) [inline src/Airship/Types.hs]: I think we removed our use of it, too. Definitely happy to get rid of it.

Comment (tmcgilchrist) [inline src/Airship/Types.hs]: https://github.com/helium/airship/pull/98

File: src/Airship/Internal/Helpers.hs [hunk 15]
```diff
@@ -94,21 +97,74 @@ toWaiResponse Response{..} cfg trace quip =
...
+    | (HTTP.statusIsClientError _responseStatus ||
+       HTTP.statusIsServerError _responseStatus) &&
+       isResponseBodyEmpty _responseBody = do
+           req <- request
+           let reqHeaders = requestHeaders req
+               acceptStr = lookup HTTP.hAccept reqHeaders
+               errBodies = map dupContentType <$> M.lookup _responseStatus errResps
+               mResp = join $ mapAcceptMedia <$> errBodies <*> acceptStr
+           when (isJust mResp) $ do
+               let (Just (ct, body)) = mResp
```
Comment (charleso) [inline src/Airship/Internal/Helpers.hs]: Just a nit reminder, but would definitely prefer to use `forM` here rather than having to use an unsafe cast.

``` haskell
forM_ mResp $ \(ct, body) -> do
  putResponseBody =<< body
  addResponseHeader ("Content-Type", renderHeader ct)
```

Comment (kellymclaughlin) [inline src/Airship/Internal/Helpers.hs]: Thanks for the reminder! I just forgot about that bit of feedback. Will change.

File: src/Airship/Internal/Helpers.hs [hunk 16]
```diff
@@ -94,21 +98,73 @@ toWaiResponse Response{..} cfg trace quip =
...
+            let default404Resp = Response HTTP.status404 [(HTTP.hContentType, "text/plain")] Empty
+            respond =<< run req (do
+                (response, trace) <- eitherResponse nowTime mempty [] req (errorResponse errors default404Resp)
+                return $ toWaiResponse response cfg (traceHeader trace) quip)
+        Just (resource, (params', matched)) -> do
+            -- Merge the application-wide error responses with the resource
+            -- specific error responses favoring the resource-specific
+            --in case of conflict
+            let mergedResource = resource { errorResponses = M.union (errorResponses resource) errors }
+            respond =<< run req (do
```
Comment (charleso) [inline src/Airship/Internal/Helpers.hs]: Vaguely wondering if it's worth removing some of the dupe here? Something along the lines of. EDIT: Realised you don't need to update the `errorResponses` on the resource any more 

``` haskell
let (er, (params, matched), r) = case route routeMapping pInfo of
  Nothing ->
    (errors, (mempty, []), return $ Response HTTP.status404 [(HTTP.hContentType, "text/plain")] Empty)
  Just (resource, pm) ->
    (M.union (errorResponses resource) errors, pm, flow resource)
respond =<< run req (do
  (response, trace) <- eitherResponse nowTime params matched req (r >>= errorResponse er)
  return $ toWaiResponse response cfg (traceHeader trace) quip)
```

**Comment by charleso** [comment]
:+1: Nice work Kelly.

File: src/Airship/Internal/Helpers.hs [hunk 17]
```diff
@@ -94,21 +98,73 @@ toWaiResponse Response{..} cfg trace quip =
...
+                (response, trace) <- eitherResponse nowTime params' matched req (flow mergedResource >>= errorResponse (errorResponses mergedResource))
+                return $ toWaiResponse response cfg (traceHeader trace) quip)
+
+
+-- | If the Response body is Empty the response body is set based on the error responses
+-- | provided by the application and resource. If the response body is not Empty or
+-- | there are no error response configured for the status code in the Response then no
+-- | action is taken. The contents of the 'Webmachine'' response body will be streamed
```
Comment (charleso) [inline src/Airship/Internal/Helpers.hs]: Super nit, but the extra `|` on lines 2+ doesn't seem right. Not sure what haddock would do with them.

File: src/Airship/Internal/Route.hs (hunk 2 above)
Comment (kellymclaughlin) [inline src/Airship/Internal/Route.hs]: I've created a simple type alias, `MatchedRoute`, that reduces some of the line noise, but I'm afraid it doesn't do much to increase clarity. I think doing anything more, however, goes beyond the scope of this PR and is better handled separately.
